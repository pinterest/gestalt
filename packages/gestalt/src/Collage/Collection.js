// @flow strict
/*
  # Collection

  Collection is component that can absolutely position and virtualize items. It's meant to be an exteremly simple workhorse to re-implement custom layouts. It's not perfect for all situations but it does have a closed API and is pretty performant.

  ## How it works

  There are two central concepts to Collection - the content layer and the viewport layer.

  Content layer               Viewport                   Composite layer

  +----------------+          +----------------+          +----------------+
  |                |          |                |          |                |
  | +--+ +--+ +--+ |          |                |          |                |
  | |  | |  | |  | |          |                |          |                |
  | |  | |  | |  | |          | +------------+ |          | +------------+ |
  | +--+ |  | |  | |          | |            | |          | |--+ |  | |  | |
  |      |  | +--+ |          | |            | |          | |    |  | +--| |
  | +--+ +--+      |          | |            | |          | |--+ +--+    | |
  | |  |      +--+ | +------> | |            | | +------> | |  |      +--| |
  | +--+ +--+ |  | |          | |            | |          | |--+ +--+ |  | |
  |      |  | |  | |          | |            | |          | |    |  | |  | |
  | +--+ |  | +--+ |          | |            | |          | |--+ |  | +--| |
  | |  | |  |      |          | |            | |          | |  | |  |    | |
  | |  | |  | +--+ |          | +------------+ |          | +------------+ |
  | |  | |  | |  | |          |                |          |                |
  | +--+ +--+ +--+ |          |                |          |                |
  |                |          |                |          |                |
  +----------------+          +----------------+          +----------------+

  You position your items in the virtual content layer, then you define the viewport layer (by default this is the full size of the content layer). Collection then takes those items, and the viewport rendering the content layer with only the items that fit inside the viewport.

  ## Assumptions

  1. Your layout is pre-calculated. Collection expects layout to be passed in as a property. If your layout isn't ready (i.e. something needs to be measured) the caller needs to figure that out.
  2. Passed in Item are somewhat pure - they can be mounted and unmounted at anytime so Item's lifecycle methods need to account for that.
  3. Passed in Items are bound to their data-source seperately. All they receive is an index, so they need to have any external data sources in scope.
  4. The viewport can be any size. Most windowing/recycling solutions implement some sort of overscanning, however Collection leaves this up the the parent.

*/
import { type Node } from 'react';
import layoutStyles from '../Layout.css';

type Props = {
  Item?: ({ idx: number }) => Node,
  layout: $ReadOnlyArray<{
    top: number,
    left: number,
    width: number,
    height: number,
  }>,
  renderItem?: ({ idx: number }) => Node,
  viewportTop?: number,
  viewportLeft?: number,
  viewportWidth?: number,
  viewportHeight?: number,
};

export default function Collection(props: Props): Node {
  const { Item, layout = [], renderItem, viewportTop = 0, viewportLeft = 0 } = props;

  // Calculate the full dimensions of the item layer
  const width = Math.max(...layout.map((item) => item.left + item.width));
  const height = Math.max(...layout.map((item) => item.top + item.height));

  // Default the viewport to being the full width of the content layer
  const { viewportWidth = width, viewportHeight = height } = props;

  // Calculates which items from the item layer to render in the viewport
  // layer.
  const items = layout.reduce(
    (
      acc: $ReadOnlyArray<{
        height: number,
        idx: number,
        left: number,
        top: number,
        width: number,
      }>,
      position,
      idx,
    ) => {
      const newAcc = [...acc];
      if (
        position.top + position.height > viewportTop &&
        position.top < viewportHeight + viewportTop &&
        position.left < viewportWidth + viewportLeft &&
        position.left + position.width > viewportLeft
      ) {
        newAcc.push({ idx, ...position });
      }
      return newAcc;
    },
    [],
  );

  return (
    <div className={layoutStyles.relative} style={{ width, height }}>
      {items.map(({ idx, ...style }) => (
        <div key={idx} className={layoutStyles.absolute} style={style}>
          {renderItem ? renderItem({ idx }) : Item && <Item idx={idx} />}
        </div>
      ))}
    </div>
  );
}
