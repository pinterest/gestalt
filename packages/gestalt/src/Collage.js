// @flow
import * as React from 'react';
import PropTypes from 'prop-types';
import Collection from './Collection.js';

type Column = 2 | 3 | 4;

/*

This function just implements the design constraints for asymetrical columns in
a very simple way. It's not perfect, but it's very predictible. If you do
anything, improve it so that it takes these constraints and generates them
for n > 4.

1 + n columns:
  first column is 50/50
  can't have 3 columns in a row be 50/50

*/
const UP = [0.75, 0.25];
const DOWN = [0.25, 0.75];
const MID = [0.5, 0.5];
const columnLayout = (numOfColumns: 1 | Column) => {
  switch (numOfColumns) {
    case 1:
      return [[MID], [UP], [DOWN]];
    case 2:
      return [[UP, MID], [DOWN, MID], [MID, UP], [MID, DOWN]];
    case 4:
      return [
        [MID, UP, MID, MID],
        [MID, DOWN, MID, MID],
        [MID, MID, UP, MID],
        [MID, MID, DOWN, MID],
        [MID, UP, MID, DOWN],
        [MID, DOWN, MID, UP],
      ];
    case 3:
    default:
      return [
        [MID, UP, MID],
        [MID, DOWN, MID],
        [MID, MID, UP],
        [MID, MID, DOWN],
      ];
  }
};

const paddingAll = (gutter, positions) =>
  positions.map(({ top, left, width, height }) => ({
    top,
    left,
    width: width - gutter,
    height: height - gutter,
  }));

const mindex = arr =>
  arr.reduce((minIndex, item, i) => (item < arr[minIndex] ? i : minIndex), 0);

const columnsForCollageWithCover = (numOfColumns: Column) =>
  numOfColumns === 4 ? 2 : 1;

function getCollageLayout({
  gutter,
  cover,
  columns: numCols,
  height: h,
  width: w,
  layoutKey,
}: {
  gutter: number,
  cover: boolean,
  columns: Column,
  height: number,
  width: number,
  layoutKey: number,
}) {
  let positions = [];
  const width = w + gutter;
  const height = h + gutter;

  // If there's a cover image, we'll add that later. It should be a little
  // less than half the width of the collage. We do this now (and not later
  // when we add the cover image) because of `columnLayout`'s constraints
  // needing the exact number of columns that are displayed.
  const gridCols = cover ? columnsForCollageWithCover(numCols) : numCols;

  // Selects the layout that we're going to use for the grid
  const columns = columnLayout(gridCols);
  const layoutIdx = layoutKey % columns.length;
  const layout = columns[layoutIdx];

  // This does a really simple version of our masonry layout. Why replicate
  // that here? a.) fewer dependencies and b.) we need the algorithm to
  // roughly preserve the order of the collages from when they were ordered
  // by Masonry.
  const colHeights = new Array(gridCols).fill(0);
  const colCounts = new Array(gridCols).fill(0);

  // We iterate over every position that we think we could _potentially_ have
  // so that we can fill them with empty sections if need be.
  for (let i = 0; i < 2 * gridCols; i += 1) {
    const col = mindex(colHeights);
    const colIdx = colCounts[col];
    const itemHeight = layout[col][colIdx] * height;

    positions.push({
      top: colHeights[col],
      left: col * (width / numCols),
      width: width / numCols,
      height: itemHeight,
    });

    colHeights[col] += itemHeight;
    colCounts[col] += 1;
  }

  // If we have a cover image, figure out how big it is, then move all the
  // existing columns over.
  if (cover) {
    const coverImageWidth = Math.ceil(numCols / 2) * (width / numCols);
    positions = positions.map(position => ({
      ...position,
      left: coverImageWidth + position.left,
    }));
    positions.unshift({ top: 0, left: 0, width: coverImageWidth, height });
  }

  // This adds the space between any items that we have. It's nice to do
  // this as a separate pass after everything else, because the math earlier
  // becomes easier and it's less brittle to change.
  if (gutter) {
    positions = paddingAll(gutter, positions);
  }

  return positions;
}

type Props = {|
  columns: Column,
  cover?: boolean,
  gutter?: number,
  height: number,
  layoutKey?: number,
  renderImage: ({|
    width: number,
    height: number,
    index: number,
  |}) => React.Node,
  width: number,
|};

export default function Collage(props: Props) {
  const {
    columns,
    cover,
    gutter,
    height,
    layoutKey,
    renderImage,
    width,
  } = props;
  const positions = getCollageLayout({
    columns,
    cover: !!cover,
    width,
    height,
    gutter: gutter || 0,
    layoutKey: layoutKey || 0,
  });
  return (
    <Collection
      Item={({ idx: index }) =>
        renderImage({
          index,
          width: positions[index].width,
          height: positions[index].height,
        })
      }
      layout={positions}
    />
  );
}

Collage.propTypes = {
  columns: PropTypes.oneOf([2, 3, 4]).isRequired,
  cover: PropTypes.bool,
  gutter: PropTypes.number,
  height: PropTypes.number.isRequired,
  layoutKey: PropTypes.number,
  renderImage: PropTypes.func.isRequired,
  width: PropTypes.number.isRequired,
};
