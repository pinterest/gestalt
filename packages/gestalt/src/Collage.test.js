// // @flow
import * as React from 'react';
import { create } from 'react-test-renderer';
import Collage from './Collage.js';

type CollageImageProps = {
  width: number,
  height: number,
  index: number,
};

describe('<Collage />', () => {
  function CollageImage({ width, height, index }: CollageImageProps) {
    return <div style={{ width, height }}>{index}</div>;
  }

  const collageProps = {
    columns: 2,
    gutter: 2,
    height: 200,
    width: 300,
    renderImage: CollageImage,
  };

  it('should match the snapshot', () => {
    expect(create(<Collage {...collageProps} />).toJSON()).toMatchSnapshot();
  });

  it('should match the snapshot for 3 columns', () => {
    expect(
      create(<Collage {...collageProps} columns={3} />).toJSON()
    ).toMatchSnapshot();
  });

  it('should match the snapshot for 4 columns', () => {
    expect(
      create(<Collage {...collageProps} columns={4} />).toJSON()
    ).toMatchSnapshot();
  });

  it('should match the snapshot for 4 columns with a custom layout key', () => {
    expect(
      create(<Collage {...collageProps} columns={4} layoutKey={1} />).toJSON()
    ).toMatchSnapshot();
  });

  it('should match the snapshot when cover image is used', () => {
    expect(
      create(<Collage {...collageProps} cover />).toJSON()
    ).toMatchSnapshot();
  });
});
