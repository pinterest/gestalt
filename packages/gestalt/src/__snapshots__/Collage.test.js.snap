// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Collage /> should match the snapshot for 0 images 1`] = `
<div
  className="relative"
  style={
    Object {
      "height": 200,
      "width": 300,
    }
  }
>
  <div
    className="absolute"
    style={
      Object {
        "height": 150,
        "left": 0,
        "top": 0,
        "width": 150,
      }
    }
  >
    <div
      className="box lightGrayBg"
      style={
        Object {
          "height": 150,
          "width": 150,
        }
      }
    />
  </div>
  <div
    className="absolute"
    style={
      Object {
        "height": 100,
        "left": 150,
        "top": 0,
        "width": 150,
      }
    }
  >
    <div
      className="box lightGrayBg"
      style={
        Object {
          "height": 100,
          "width": 150,
        }
      }
    />
  </div>
  <div
    className="absolute"
    style={
      Object {
        "height": 100,
        "left": 150,
        "top": 100,
        "width": 150,
      }
    }
  >
    <div
      className="box lightGrayBg"
      style={
        Object {
          "height": 100,
          "width": 150,
        }
      }
    />
  </div>
  <div
    className="absolute"
    style={
      Object {
        "height": 50,
        "left": 0,
        "top": 150,
        "width": 150,
      }
    }
  >
    <div
      className="box lightGrayBg"
      style={
        Object {
          "height": 50,
          "width": 150,
        }
      }
    />
  </div>
</div>
`;

exports[`<Collage /> should match the snapshot for 1 image 1`] = `
<div
  className="relative"
  style={
    Object {
      "height": 200,
      "width": 300,
    }
  }
>
  <div
    className="absolute"
    style={
      Object {
        "height": 100,
        "left": 0,
        "top": 0,
        "width": 150,
      }
    }
  >
    <div
      className="Mask square"
      style={
        Object {
          "height": 100,
          "width": undefined,
        }
      }
    >
      <div
        aria-label="collage"
        className="cover"
        role="img"
        style={
          Object {
            "backgroundColor": "#FCDB9B",
            "backgroundImage": "url('foo.png')",
          }
        }
      />
      <div
        className="wash"
      />
    </div>
  </div>
  <div
    className="absolute"
    style={
      Object {
        "height": 50,
        "left": 150,
        "top": 0,
        "width": 150,
      }
    }
  >
    <div
      className="box lightGrayBg"
      style={
        Object {
          "height": 50,
          "width": 150,
        }
      }
    />
  </div>
  <div
    className="absolute"
    style={
      Object {
        "height": 150,
        "left": 150,
        "top": 50,
        "width": 150,
      }
    }
  >
    <div
      className="box lightGrayBg"
      style={
        Object {
          "height": 150,
          "width": 150,
        }
      }
    />
  </div>
  <div
    className="absolute"
    style={
      Object {
        "height": 100,
        "left": 0,
        "top": 100,
        "width": 150,
      }
    }
  >
    <div
      className="box lightGrayBg"
      style={
        Object {
          "height": 100,
          "width": 150,
        }
      }
    />
  </div>
</div>
`;

exports[`<Collage /> should match the snapshot for 2 images 1`] = `
<div
  className="relative"
  style={
    Object {
      "height": 200,
      "width": 300,
    }
  }
>
  <div
    className="absolute"
    style={
      Object {
        "height": 150,
        "left": 0,
        "top": 0,
        "width": 150,
      }
    }
  >
    <div
      className="Mask square"
      style={
        Object {
          "height": 150,
          "width": undefined,
        }
      }
    >
      <div
        aria-label="collage"
        className="cover"
        role="img"
        style={
          Object {
            "backgroundColor": "#FCDB9B",
            "backgroundImage": "url('foo.png')",
          }
        }
      />
      <div
        className="wash"
      />
    </div>
  </div>
  <div
    className="absolute"
    style={
      Object {
        "height": 100,
        "left": 150,
        "top": 0,
        "width": 150,
      }
    }
  >
    <div
      className="Mask square"
      style={
        Object {
          "height": 100,
          "width": undefined,
        }
      }
    >
      <div
        aria-label="collage"
        className="cover"
        role="img"
        style={
          Object {
            "backgroundColor": "#FCDB9B",
            "backgroundImage": "url('foo.png')",
          }
        }
      />
      <div
        className="wash"
      />
    </div>
  </div>
  <div
    className="absolute"
    style={
      Object {
        "height": 100,
        "left": 150,
        "top": 100,
        "width": 150,
      }
    }
  >
    <div
      className="box lightGrayBg"
      style={
        Object {
          "height": 100,
          "width": 150,
        }
      }
    />
  </div>
  <div
    className="absolute"
    style={
      Object {
        "height": 50,
        "left": 0,
        "top": 150,
        "width": 150,
      }
    }
  >
    <div
      className="box lightGrayBg"
      style={
        Object {
          "height": 50,
          "width": 150,
        }
      }
    />
  </div>
</div>
`;

exports[`<Collage /> should match the snapshot for 100 images and 4 columns 1`] = `
<div
  className="relative"
  style={
    Object {
      "height": 200,
      "width": 300,
    }
  }
>
  <div
    className="absolute"
    style={
      Object {
        "height": 100,
        "left": 0,
        "top": 0,
        "width": 75,
      }
    }
  >
    <div
      className="Mask square"
      style={
        Object {
          "height": 100,
          "width": undefined,
        }
      }
    >
      <div
        aria-label="collage"
        className="cover"
        role="img"
        style={
          Object {
            "backgroundColor": "#FCDB9B",
            "backgroundImage": "url('foo.png')",
          }
        }
      />
      <div
        className="wash"
      />
    </div>
  </div>
  <div
    className="absolute"
    style={
      Object {
        "height": 150,
        "left": 75,
        "top": 0,
        "width": 75,
      }
    }
  >
    <div
      className="Mask square"
      style={
        Object {
          "height": 150,
          "width": undefined,
        }
      }
    >
      <div
        aria-label="collage"
        className="cover"
        role="img"
        style={
          Object {
            "backgroundColor": "#FCDB9B",
            "backgroundImage": "url('foo.png')",
          }
        }
      />
      <div
        className="wash"
      />
    </div>
  </div>
  <div
    className="absolute"
    style={
      Object {
        "height": 100,
        "left": 150,
        "top": 0,
        "width": 75,
      }
    }
  >
    <div
      className="Mask square"
      style={
        Object {
          "height": 100,
          "width": undefined,
        }
      }
    >
      <div
        aria-label="collage"
        className="cover"
        role="img"
        style={
          Object {
            "backgroundColor": "#FCDB9B",
            "backgroundImage": "url('foo.png')",
          }
        }
      />
      <div
        className="wash"
      />
    </div>
  </div>
  <div
    className="absolute"
    style={
      Object {
        "height": 50,
        "left": 225,
        "top": 0,
        "width": 75,
      }
    }
  >
    <div
      className="Mask square"
      style={
        Object {
          "height": 50,
          "width": undefined,
        }
      }
    >
      <div
        aria-label="collage"
        className="cover"
        role="img"
        style={
          Object {
            "backgroundColor": "#FCDB9B",
            "backgroundImage": "url('foo.png')",
          }
        }
      />
      <div
        className="wash"
      />
    </div>
  </div>
  <div
    className="absolute"
    style={
      Object {
        "height": 150,
        "left": 225,
        "top": 50,
        "width": 75,
      }
    }
  >
    <div
      className="Mask square"
      style={
        Object {
          "height": 150,
          "width": undefined,
        }
      }
    >
      <div
        aria-label="collage"
        className="cover"
        role="img"
        style={
          Object {
            "backgroundColor": "#FCDB9B",
            "backgroundImage": "url('foo.png')",
          }
        }
      />
      <div
        className="wash"
      />
    </div>
  </div>
  <div
    className="absolute"
    style={
      Object {
        "height": 100,
        "left": 0,
        "top": 100,
        "width": 75,
      }
    }
  >
    <div
      className="Mask square"
      style={
        Object {
          "height": 100,
          "width": undefined,
        }
      }
    >
      <div
        aria-label="collage"
        className="cover"
        role="img"
        style={
          Object {
            "backgroundColor": "#FCDB9B",
            "backgroundImage": "url('foo.png')",
          }
        }
      />
      <div
        className="wash"
      />
    </div>
  </div>
  <div
    className="absolute"
    style={
      Object {
        "height": 100,
        "left": 150,
        "top": 100,
        "width": 75,
      }
    }
  >
    <div
      className="Mask square"
      style={
        Object {
          "height": 100,
          "width": undefined,
        }
      }
    >
      <div
        aria-label="collage"
        className="cover"
        role="img"
        style={
          Object {
            "backgroundColor": "#FCDB9B",
            "backgroundImage": "url('foo.png')",
          }
        }
      />
      <div
        className="wash"
      />
    </div>
  </div>
  <div
    className="absolute"
    style={
      Object {
        "height": 50,
        "left": 75,
        "top": 150,
        "width": 75,
      }
    }
  >
    <div
      className="Mask square"
      style={
        Object {
          "height": 50,
          "width": undefined,
        }
      }
    >
      <div
        aria-label="collage"
        className="cover"
        role="img"
        style={
          Object {
            "backgroundColor": "#FCDB9B",
            "backgroundImage": "url('foo.png')",
          }
        }
      />
      <div
        className="wash"
      />
    </div>
  </div>
</div>
`;

exports[`<Collage /> should match the snapshot when cover image is used 1`] = `
<div
  className="relative"
  style={
    Object {
      "height": 200,
      "width": 300,
    }
  }
>
  <div
    className="absolute"
    style={
      Object {
        "height": 200,
        "left": 0,
        "top": 0,
        "width": 150,
      }
    }
  >
    <div
      className="Mask square"
      style={
        Object {
          "height": 200,
          "width": undefined,
        }
      }
    >
      <div
        style={
          Object {
            "height": 200,
            "width": 150,
          }
        }
      />
    </div>
  </div>
  <div
    className="absolute"
    style={
      Object {
        "height": 150,
        "left": 150,
        "top": 0,
        "width": 150,
      }
    }
  >
    <div
      className="Mask square"
      style={
        Object {
          "height": 150,
          "width": undefined,
        }
      }
    >
      <div
        aria-label="collage"
        className="cover"
        role="img"
        style={
          Object {
            "backgroundColor": "#FCDB9B",
            "backgroundImage": "url('foo.png')",
          }
        }
      />
      <div
        className="wash"
      />
    </div>
  </div>
  <div
    className="absolute"
    style={
      Object {
        "height": 50,
        "left": 150,
        "top": 150,
        "width": 150,
      }
    }
  >
    <div
      className="Mask square"
      style={
        Object {
          "height": 50,
          "width": undefined,
        }
      }
    >
      <div
        aria-label="collage"
        className="cover"
        role="img"
        style={
          Object {
            "backgroundColor": "#FCDB9B",
            "backgroundImage": "url('foo.png')",
          }
        }
      />
      <div
        className="wash"
      />
    </div>
  </div>
</div>
`;

exports[`<Collage /> should match the snapshot when imagesAreRounded is true 1`] = `
<div
  className="relative"
  style={
    Object {
      "height": 200,
      "width": 300,
    }
  }
>
  <div
    className="absolute"
    style={
      Object {
        "height": 150,
        "left": 0,
        "top": 0,
        "width": 150,
      }
    }
  >
    <div
      className="Mask rounded"
      style={
        Object {
          "height": 150,
          "width": undefined,
        }
      }
    >
      <div
        aria-label="collage"
        className="cover"
        role="img"
        style={
          Object {
            "backgroundColor": "#FCDB9B",
            "backgroundImage": "url('foo.png')",
          }
        }
      />
      <div
        className="wash"
      />
    </div>
  </div>
  <div
    className="absolute"
    style={
      Object {
        "height": 100,
        "left": 150,
        "top": 0,
        "width": 150,
      }
    }
  >
    <div
      className="Mask rounded"
      style={
        Object {
          "height": 100,
          "width": undefined,
        }
      }
    >
      <div
        aria-label="collage"
        className="cover"
        role="img"
        style={
          Object {
            "backgroundColor": "#FCDB9B",
            "backgroundImage": "url('foo.png')",
          }
        }
      />
      <div
        className="wash"
      />
    </div>
  </div>
  <div
    className="absolute"
    style={
      Object {
        "height": 100,
        "left": 150,
        "top": 100,
        "width": 150,
      }
    }
  >
    <div
      className="Mask rounded"
      style={
        Object {
          "height": 100,
          "width": undefined,
        }
      }
    >
      <div
        aria-label="collage"
        className="cover"
        role="img"
        style={
          Object {
            "backgroundColor": "#FCDB9B",
            "backgroundImage": "url('foo.png')",
          }
        }
      />
      <div
        className="wash"
      />
    </div>
  </div>
  <div
    className="absolute"
    style={
      Object {
        "height": 50,
        "left": 0,
        "top": 150,
        "width": 150,
      }
    }
  >
    <div
      className="Mask rounded"
      style={
        Object {
          "height": 50,
          "width": undefined,
        }
      }
    >
      <div
        aria-label="collage"
        className="cover"
        role="img"
        style={
          Object {
            "backgroundColor": "#FCDB9B",
            "backgroundImage": "url('foo.png')",
          }
        }
      />
      <div
        className="wash"
      />
    </div>
  </div>
</div>
`;

exports[`<Collage /> should match the snapshot when one of the images has width=0 and height=0 1`] = `
<div
  className="relative"
  style={
    Object {
      "height": 200,
      "width": 300,
    }
  }
>
  <div
    className="absolute"
    style={
      Object {
        "height": 150,
        "left": 0,
        "top": 0,
        "width": 150,
      }
    }
  >
    <div
      className="Mask square"
      style={
        Object {
          "height": 150,
          "width": undefined,
        }
      }
    >
      <div
        aria-label="collage"
        className="cover"
        role="img"
        style={
          Object {
            "backgroundColor": "#FCDB9B",
            "backgroundImage": "url('foo.png')",
          }
        }
      />
      <div
        className="wash"
      />
    </div>
  </div>
  <div
    className="absolute"
    style={
      Object {
        "height": 100,
        "left": 150,
        "top": 0,
        "width": 150,
      }
    }
  >
    <div
      className="Mask square"
      style={
        Object {
          "height": 100,
          "width": undefined,
        }
      }
    >
      <div
        aria-label="collage"
        className="cover"
        role="img"
        style={
          Object {
            "backgroundColor": "#FCDB9B",
            "backgroundImage": "url('foo.png')",
          }
        }
      />
      <div
        className="wash"
      />
    </div>
  </div>
  <div
    className="absolute"
    style={
      Object {
        "height": 100,
        "left": 150,
        "top": 100,
        "width": 150,
      }
    }
  >
    <div
      className="Mask square"
      style={
        Object {
          "height": 100,
          "width": undefined,
        }
      }
    >
      <div
        aria-label="collage"
        className="cover"
        role="img"
        style={
          Object {
            "backgroundColor": "#FCDB9B",
            "backgroundImage": "url('foo.png')",
          }
        }
      />
      <div
        className="wash"
      />
    </div>
  </div>
  <div
    className="absolute"
    style={
      Object {
        "height": 50,
        "left": 0,
        "top": 150,
        "width": 150,
      }
    }
  >
    <div
      className="Mask square"
      style={
        Object {
          "height": 50,
          "width": undefined,
        }
      }
    >
      <div
        aria-label="collage"
        className="cover"
        role="img"
        style={
          Object {
            "backgroundColor": "#FCDB9B",
            "backgroundImage": "url('foo.png')",
          }
        }
      />
      <div
        className="wash"
      />
    </div>
  </div>
</div>
`;

exports[`<Collage /> should match the snapshot when spacing (around) 1`] = `
<div
  className="relative"
  style={
    Object {
      "height": 200,
      "width": 300,
    }
  }
>
  <div
    className="absolute"
    style={
      Object {
        "height": 146,
        "left": 2,
        "top": 2,
        "width": 146,
      }
    }
  >
    <div
      className="Mask rounded"
      style={
        Object {
          "height": 146,
          "width": undefined,
        }
      }
    >
      <div
        aria-label="collage"
        className="cover"
        role="img"
        style={
          Object {
            "backgroundColor": "#FCDB9B",
            "backgroundImage": "url('foo.png')",
          }
        }
      />
      <div
        className="wash"
      />
    </div>
  </div>
  <div
    className="absolute"
    style={
      Object {
        "height": 96,
        "left": 152,
        "top": 2,
        "width": 146,
      }
    }
  >
    <div
      className="Mask rounded"
      style={
        Object {
          "height": 96,
          "width": undefined,
        }
      }
    >
      <div
        aria-label="collage"
        className="cover"
        role="img"
        style={
          Object {
            "backgroundColor": "#FCDB9B",
            "backgroundImage": "url('foo.png')",
          }
        }
      />
      <div
        className="wash"
      />
    </div>
  </div>
  <div
    className="absolute"
    style={
      Object {
        "height": 96,
        "left": 152,
        "top": 102,
        "width": 146,
      }
    }
  >
    <div
      className="Mask rounded"
      style={
        Object {
          "height": 96,
          "width": undefined,
        }
      }
    >
      <div
        aria-label="collage"
        className="cover"
        role="img"
        style={
          Object {
            "backgroundColor": "#FCDB9B",
            "backgroundImage": "url('foo.png')",
          }
        }
      />
      <div
        className="wash"
      />
    </div>
  </div>
  <div
    className="absolute"
    style={
      Object {
        "height": 46,
        "left": 2,
        "top": 152,
        "width": 146,
      }
    }
  >
    <div
      className="Mask rounded"
      style={
        Object {
          "height": 46,
          "width": undefined,
        }
      }
    >
      <div
        aria-label="collage"
        className="cover"
        role="img"
        style={
          Object {
            "backgroundColor": "#FCDB9B",
            "backgroundImage": "url('foo.png')",
          }
        }
      />
      <div
        className="wash"
      />
    </div>
  </div>
</div>
`;

exports[`<Collage /> should match the snapshot when spacing (between) 1`] = `
<div
  className="relative"
  style={
    Object {
      "height": 202,
      "width": 302,
    }
  }
>
  <div
    className="absolute"
    style={
      Object {
        "height": 149,
        "left": 0,
        "top": 0,
        "width": 148,
      }
    }
  >
    <div
      className="Mask rounded"
      style={
        Object {
          "height": 149,
          "width": undefined,
        }
      }
    >
      <div
        aria-label="collage"
        className="cover"
        role="img"
        style={
          Object {
            "backgroundColor": "#FCDB9B",
            "backgroundImage": "url('foo.png')",
          }
        }
      />
      <div
        className="wash"
      />
    </div>
  </div>
  <div
    className="absolute"
    style={
      Object {
        "height": 98,
        "left": 152,
        "top": 0,
        "width": 148,
      }
    }
  >
    <div
      className="Mask rounded"
      style={
        Object {
          "height": 98,
          "width": undefined,
        }
      }
    >
      <div
        aria-label="collage"
        className="cover"
        role="img"
        style={
          Object {
            "backgroundColor": "#FCDB9B",
            "backgroundImage": "url('foo.png')",
          }
        }
      />
      <div
        className="wash"
      />
    </div>
  </div>
  <div
    className="absolute"
    style={
      Object {
        "height": 98,
        "left": 152,
        "top": 102,
        "width": 148,
      }
    }
  >
    <div
      className="Mask rounded"
      style={
        Object {
          "height": 98,
          "width": undefined,
        }
      }
    >
      <div
        aria-label="collage"
        className="cover"
        role="img"
        style={
          Object {
            "backgroundColor": "#FCDB9B",
            "backgroundImage": "url('foo.png')",
          }
        }
      />
      <div
        className="wash"
      />
    </div>
  </div>
  <div
    className="absolute"
    style={
      Object {
        "height": 47,
        "left": 0,
        "top": 153,
        "width": 148,
      }
    }
  >
    <div
      className="Mask rounded"
      style={
        Object {
          "height": 47,
          "width": undefined,
        }
      }
    >
      <div
        aria-label="collage"
        className="cover"
        role="img"
        style={
          Object {
            "backgroundColor": "#FCDB9B",
            "backgroundImage": "url('foo.png')",
          }
        }
      />
      <div
        className="wash"
      />
    </div>
  </div>
</div>
`;
